crates/ruff_cli/src/cache.rs

```diff
@@ -442,52 +449,98 @@ mod tests {
+
+        cache.store().unwrap();
+    }
+
+    #[test]
+    fn cache_adds_files_on_lint() {
+        let source: &[u8] = b"a = 1\n\n__all__ = list([\"a\", \"b\"])\n";
+
+        let test_cache = TestCache::new("cache_adds_files_on_lint");
+        let cache = test_cache.open();
+        test_cache.write_source_file("source_1.py", source);
+        test_cache.write_source_file("source_2.py", source);
         assert_eq!(cache.new_files.lock().unwrap().len(), 0);
 
-        let path = package_root.join("source.py");
-        let mut expected_diagnostics = lint_path(
-            &path,
-            Some(&package_root),
-            &settings,
-            Some(&cache),
-            flags::Noqa::Enabled,
-            flags::FixMode::Generate,
-        )
-        .unwrap();
-        assert_eq!(cache.new_files.lock().unwrap().len(), 1);
+        cache.store().unwrap();
+        let cache = test_cache.open();
+
+        test_cache
+            .lint_file_with_cache("source_1.py", &cache)
+            .expect("Failed to lint test file");
+        test_cache
+            .lint_file_with_cache("source_2.py", &cache)
+            .expect("Failed to lint test file");
+        assert_eq!(
+            cache.new_files.lock().unwrap().len(),
+            2,
+            "Both files should be added to the cache"
+        );
 
         cache.store().unwrap();
+    }
 
-        let tests = [
-            // File change.
-            (|path| {
-                let mut file = fs::OpenOptions::new()
-                    .write(true)
-                    .truncate(true)
-                    .open(path)?;
-                file.write_all(SOURCE)?;
-                file.sync_data()?;
-                Ok(|_| Ok(()))
-            }) as fn(&Path) -> io::Result<fn(&Path) -> io::Result<()>>,
-            // Regression for issue #3086.
-            #[cfg(unix)]
-            |path| {
-                flip_execute_permission_bit(path)?;
-                Ok(flip_execute_permission_bit)
-            },
-        ];
+    #[test]
+    fn cache_invalidated_on_file_modified_time() {
+        let source: &[u8] = b"a = 1\n\n__all__ = list([\"a\", \"b\"])\n";
+
+        let test_cache = TestCache::new("cache_invalidated_on_file_modified_time");
+        let cache = test_cache.open();
+        test_cache.write_source_file("source.py", source);
+        assert_eq!(cache.new_files.lock().unwrap().len(), 0);
+
+        let expected_diagnostics = test_cache
+            .lint_file_with_cache("source.py", &cache)
+            .expect("Failed to lint test file");
+
+        cache.store().unwrap();
+        let cache = test_cache.open();
+
+        // Write the same contents to the source file (updating the modified time)
+        // We must sleep for a little or this can happen too fast on some systems resulting in a modified time collision.
+        // Once `std::fs::File.set_modified` is stable we can use that instead.
+        std::thread::sleep(std::time::Duration::from_millis(10));
```

Comment:
The answer might be no, but feel obligated to ask if there’s any way to design around this potential test flakiness.

Reply:
Thanks for the additional nudge, testing https://github.com/astral-sh/ruff/pull/5901/commits/dc322440f1277c09b2b191918806e56f2b173a54

Reply:
I don't think so. The problem here seems to be the precision at which the file system (and kernel) represent times

https://stackoverflow.com/questions/14392975/timestamp-accuracy-on-ext4-sub-millsecond

https://apenwarr.ca/log/20181113

Reply:
@MichaReiser let me know if my patch causes problems on your superspeed machine

Reply:
ohhh nice! Works perfectly